### PR TITLE
Working MCDS implementation using connectivity constraints

### LINKED ISSUE #35989: Minimum connected dominating set
### Problem Description

Currently, SageMath does not provide the ability to generate the minimum **connected** dominating set (MCDS). MCDS is an incredibly useful result for fields like telecommunications and bioinformatics, as well as for computing the Max Leaf Number parameter of a graph.

### Proposed Solution

To generate minimum connected dominating sets, we can extend the `dominating_sets` method to add Miller-Tucker-Zemlin (MTZ) connectivity constraints to the MILP, triggered by a "connected" parameter flag that defaults to false. Likewise, the `dominating_set` function will also have a parameter flag.

### Alternatives Considered

Other connectivity constraint formulations were considered (e.g. Multi-Commodity Flow), but MTZ constraints were found to be the best approach.

### Additional Information

Component: graph theory

For more information on connectivity constraints, see https://doi.org/10.1007/978-3-642-31770-5_33.

### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### PR DESCRIPTION
**Pull request by MrBanananator**
Extended `dominating_sets` to be able to generate Minimum Connected Dominating Sets using connectivity constraints. This functionality can be toggled on by setting the `connected` parameter flag to true.

Fixes #35989.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dcoudert** [comment]
Can you comment the modifications of the graph and the constraints you add. I don't have access to the pdf of the paper you use and it seems much more complicated than what it should.

**Comment by MrBanananator** [comment]
I've attached a screenshot of the MTZ constraints formulation from the paper here: https://github.com/sagemath/sage/assets/61144856/303989f4-94a2-4b94-84c5-bbc44d00c387

You can also access my breakdown/translation of the constraints here: https://docs.google.com/document/d/1zlPE6KKcjKqocCsAf-CXsYjtys6V7l5Li5sXqnOhMN8/edit?usp=sharing

**Comment by dcoudert** [comment]
Your current proposals has several issues. The main one is that it assumes that the vertices of the graph are labeled `0,1,2,...,n-1`, which is not always the case. Furthermore, it is unclear if it can be used in combination with other parameters from the method. 

I have an alternative proposal (much simpler). It suffices to add the following in `dominating_sets`, just after the case `if independent`. There is no need for a specific method to add the constraints.
```py
    if connected:
        # edges used in the spanning tree
        edge = p.new_variable(binary=True)
        # relaxed edges to test for acyclicity
        r_edge = p.new_variable(nonnegative=True)

        # 1. We want a tree
        p.add_constraint(p.sum(edge[frozenset(e)] for e in g.edge_iterator(labels=False))
                         == p.sum(b[u] for u in g) - 1)

        # 2. An edge can be in the tree if its end vertices are selected
        for u, v in g.edge_iterator(labels=False):
            p.add_constraint(edge[frozenset((u, v))] <= b[u])
            p.add_constraint(edge[frozenset((u, v))] <= b[v])

        # 3. Subtour elimination constraints from the maximum average degree
        for u, v in g.edge_iterator(labels=False):
            p.add_constraint(edge[frozenset((u, v))] <= r_edge[u, v] + r_edge[v, u])
        eps = 1 / (5 * Integer(g.order()))
        for v in g:
            p.add_constraint(p.sum(r_edge[u, v] for u in g.neighbor_iterator(v)), max=1 - eps)
```

I did some tests and its faster than with the method you propose. Try it.

**Comment by dcoudert** [comment]
To work properly with digraphs, we need the following to avoid counting twice some edge variables. Here, connectivity in digraphs is not strong connectivity. We only ensure that the underlying undirected graph is connected.
```py
    if connected:
        E = set(frozenset(e) for e in g.edge_iterator(labels=False))
        # edges used in the spanning tree
        edge = p.new_variable(binary=True, name='e')
        # relaxed edges to test for acyclicity
        r_edge = p.new_variable(nonnegative=True, name='re')

        # 1. We want a tree
        p.add_constraint(p.sum(edge[fe] for fe in E)
                         == p.sum(b[u] for u in g) - 1)

        # 2. An edge can be in the tree if its end vertices are selected
        for fe in E:
            u, v = fe
            p.add_constraint(edge[fe] <= b[u])
            p.add_constraint(edge[fe] <= b[v])

        # 3. Subtour elimination constraints
        for fe in E:
            u, v = fe
            p.add_constraint(edge[fe] <= r_edge[u, v] + r_edge[v, u])

        eps = 1 / (5 * Integer(g.order()))
        for v in g:
            p.add_constraint(p.sum(r_edge[u, v] for u in g.neighbor_iterator(v)), max=1 - eps)
```

**Comment by dcoudert** [comment]
Do you plan to implement the proposed solution ? Otherwise, I can do a PR with this solution.

**Comment by MrBanananator** [comment]
Yes, of course. I will submit it for review today. 

Interestingly, while doing some manual testing I noticed that the MTZ formulation (while outperformed on almost every other graph type) was actually more efficient than the other method for the HalfCube graph specifically. Any ideas why that may be?

**Comment by dcoudert** [comment]
> Interestingly, while doing some manual testing I noticed that the MTZ formulation (while outperformed on almost every other graph type) was actually more efficient than the other method for the HalfCube graph specifically. Any ideas why that may be?

I don't know how to explain that. It's not surprising to find an instance for which one formulation is better than another one. I already observed that for problems like the Steiner tree.

File: src/sage/graphs/domination.py [hunk 1]
```diff
@@ -261,6 +261,9 @@ def dominating_sets(g, k=1, independent=False, total=False,
     - ``total`` -- boolean (default: ``False``); when ``True``, computes total
       dominating sets (see the See the :wikipedia:`Dominating_set`)
 
+    - ``connected`` -- boolean (default: ``False``); when ``True``, computes connected
```
Comment (dcoudert) [inline src/sage/graphs/domination.py]: make sure to respect 80 columns mode for documentation

File: src/sage/graphs/domination.py [hunk 2]
```diff
@@ -338,6 +341,31 @@ def dominating_sets(g, k=1, independent=False, total=False,
...
+         [0, 1, 4, 5],
+         [0, 3, 4, 9],
+         [0, 5, 7, 8],
+         [1, 2, 3, 7],
+         [1, 6, 8, 9],
+         [2, 3, 4, 8],
+         [2, 5, 7, 9],
+         [3, 5, 6, 8],
+         [4, 6, 7, 9]]
+
```
Comment (dcoudert) [inline src/sage/graphs/domination.py:360]: Add a doctest showing that the subgraph induced by a dominating set is connected. For instance:

```
sage: all(G.subgraph(vertices=dom).is_connected() for dom in G.dominating_set(k=1, connected=True))
True
```

File: src/sage/graphs/domination.py [hunk 3]
```diff
@@ -415,8 +469,7 @@ def neighbors_iter(x):
         # Prevent finding twice a solution
         p.add_constraint(p.sum(b[u] for u in dom) <= best - 1)
 
-
-def dominating_set(g, k=1, independent=False, total=False, value_only=False,
+def dominating_set(g, k=1, independent=False, total=False, connected=False, value_only=False,
                    solver=None, verbose=0, *, integrality_tolerance=1e-3):
     r"""
     Return a minimum distance-`k` dominating set of the graph.
```
Comment (dcoudert) [inline src/sage/graphs/domination.py:481]: add description of input parameter `connected`

**Review by dcoudert — CHANGES_REQUESTED** [review]

**Comment by dcoudert** [comment]
Can you update the name of this pull request to reflect what it does. It no longer uses MTZ.

**Review by dcoudert — APPROVED** [review]
LGTM. Thanks.

**Comment by vbraun** [comment]
Docs don't build
```
[sagemath_doc_html-none] [graphs   ] The inventory files are in ../../local/share/doc/sage/inventory/en/reference/graphs.
[sagemath_doc_html-none] Error building the documentation.
[sagemath_doc_html-none] Traceback (most recent call last):
[sagemath_doc_html-none]   File "<frozen runpy>", line 198, in _run_module_as_main
[sagemath_doc_html-none]   File "<frozen runpy>", line 88, in _run_code
[sagemath_doc_html-none]   File "/home/release/Sage/src/sage_docbuild/__main__.py", line 502, in <module>
[sagemath_doc_html-none]     sys.exit(main())
[sagemath_doc_html-none]              ^^^^^^
[sagemath_doc_html-none]   File "/home/release/Sage/src/sage_docbuild/__main__.py", line 499, in main
[sagemath_doc_html-none]     builder()
[sagemath_doc_html-none]   File "/home/release/Sage/src/sage_docbuild/builders.py", line 818, in _wrapper
[sagemath_doc_html-none]     getattr(DocBuilder, build_type)(self, *args, **kwds)
[sagemath_doc_html-none]   File "/home/release/Sage/src/sage_docbuild/builders.py", line 162, in f
[sagemath_doc_html-none]     runsphinx()
[sagemath_doc_html-none]   File "/home/release/Sage/src/sage_docbuild/sphinxbuild.py", line 327, in runsphinx
[sagemath_doc_html-none]     sys.stderr.raise_errors()
[sagemath_doc_html-none]   File "/home/release/Sage/src/sage_docbuild/sphinxbuild.py", line 263, in raise_errors
[sagemath_doc_html-none]     raise OSError(self._error)
[sagemath_doc_html-none] OSError: /home/release/Sage/src/sage/graphs/domination.py:docstring of sage.graphs.domination.dominating_set:36: WARNING: Bullet list ends without a blank line; unexpected unindent.
```

File: src/sage/graphs/domination.py [hunk 4]
```diff
@@ -453,6 +512,9 @@ def dominating_set(g, k=1, independent=False, total=False, value_only=False,
     - ``total`` -- boolean (default: ``False``); when ``True``, computes a total
       dominating set (see the See the :wikipedia:`Dominating_set`)
 
+     - ``connected`` -- boolean (default: ``False``); when ``True``, computes a
+      connected dominating set (see :wikipedia:`Connected_dominating_set`)
```
Comment (dcoudert) [inline src/sage/graphs/domination.py:516]: 1 space is missing here
```diff
-      connected dominating set (see :wikipedia:`Connected_dominating_set`)
+       connected dominating set (see :wikipedia:`Connected_dominating_set`)
```

**Review by dcoudert — CHANGES_REQUESTED** [review]

**Review by dcoudert — APPROVED** [review]
LGTM.

**Pull request merged**